gcpvault.go

```diff
@@ -154,14 +187,55 @@ func PutVersionedSecrets(ctx context.Context, cfg Config, secrets map[string]int
 	return errors.Wrap(err, "unable to make vault request")
 }
 
-func checkDefaults(cfg *Config) {
+func checkDefaults(cfg *Config) error {
 	if cfg == nil {
-		return
+		return nil
+	}
+
+	if cfg.TokenCacheStorageGCS != "" && cfg.TokenCacheStorageRedis != "" {
+		return errors.New("Both Cache types are configured")
 	}
 
 	if cfg.AuthPath == "" {
 		cfg.AuthPath = "auth/gcp"
 	}
+
+	if cfg.TokenCacheStorageGCS != "" && cfg.TokenCache == nil {
+		cfg.TokenCache = TokenCacheGCS{cfg: cfg}
+	}
+
+	if cfg.TokenCacheStorageRedis != "" && cfg.TokenCache == nil {
+		cfg.TokenCache = TokenCacheRedis{cfg: cfg}
+	}
+
+	//if expiration is not set, use default
+	if cfg.CachedTokenRefreshThreshold == 0 {
+		cfg.CachedTokenRefreshThreshold = CachedTokenRefreshThresholdDefault
+	}
+
+	//if token cache timeout is not set, use default
+	if cfg.TokenCacheCtxTimeout == 0 {
+		cfg.TokenCacheCtxTimeout = TokenCacheCtxTimeoutDefault
+	}
+
+	//if the token cache name is not set, use default
+	if cfg.TokenCacheKeyName == "" {
+		cfg.TokenCacheKeyName = TokenCacheKeyNameDefault
+	}
+
+	if cfg.TokenCacheRefreshRandomOffset == 0 && cfg.CachedTokenRefreshThreshold > 0 {
+		//TOKEN_CACHE_REFRESH_RANDOM_OFFSET is not set, TOKEN_CACHE_REFRESH_THRESHOLD is set
+		refreshDuration := time.Minute * time.Duration(cfg.CachedTokenRefreshThreshold)
```

Comment:
Would it make sense to use seconds instead of minutes for the refresh threshold?  At the very least, it would eliminate the need to convert to seconds for the random offset.  (FWIW, most libraries I've come across use either seconds or milliseconds for time intervals)

Reply:
agree